### PR TITLE
Enhance user experience after pressing the signup button (#1866)

### LINKED ISSUE #1866: Bad experience in sign-up and email confirmation in Todo Template project
In sign-up by email, we send an email for confirmation to the user. In sending emails, we check for too many send emails.

The problem is:
When a user signs up by a specific email we generate a token to confirm her email. If she submits again for that email at the time, we generate another token and the previous token expires automatically, and we never send her a new email for confirmation. So, she never can confirm her email until she submits it again after a while.

### PR DESCRIPTION
**Pull request by mohabbati**
This closes #1866

### CODE REVIEW

**Review by maryamhdr — APPROVED** [review]

**Comment by msynk** [comment]
this PR needs a rewrite based on a conversation between me @mohabbati

**Comment by msynk** [comment]
@mohabbati any news or updates?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Shared/Resources/ErrorStrings.resx [hunk 1]
```diff
@@ -252,4 +252,7 @@
   <data name="WaitForResetPasswordEmailResendDelay" xml:space="preserve">
     <value>You have already requested for the reset password email for '{0}'. Try again later.</value>
   </data>
+  <data name="EmailAlreadyConfirmed" xml:space="preserve">
+    <value>Your email already confirmed.</value>
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Shared/Resources/ErrorStrings.resx:256]: The value of this message is better to be like this:
`Your email is already confirmed.`

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUpConfirmation.razor [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+<div class="page-container">
+    <div class="form--white">
+        @if (string.IsNullOrEmpty(SignUpConfirmationMessage) is false)
+        {
+            <BitMessageBar Class="form-message-bar"
+                       MessageBarType="@SignUpConfirmationMessageType"
+                       OnDismiss="() => SignUpConfirmationMessage = null">
+                @SignUpConfirmationMessage
+            </BitMessageBar>
+        }
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUpConfirmation.razor:16]: it's better to add a blank line after `@if {...}`

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUpConfirmation.razor [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        <h1 class="form-title">
+            Sign up
+        </h1>
+
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUpConfirmation.razor:20]: This blank line is redundant

**Review by maryamhdr — CHANGES_REQUESTED** [review]

**Review by msynk — CHANGES_REQUESTED** [review]

**Comment by msynk** [comment]
@mohabbati plz update this PR based on our last conversation :)

**Comment by msynk** [comment]
@mohabbati plz resolve the conflict in the `SignUp.razor` file

**Comment by msynk** [comment]
@maryamhdr plz review this PR

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/compilerconfig.json [hunk 4]
```diff
@@ -1,70 +1,70 @@
...
+    "inputFile": "Components/SignOutConfirmModal.razor.scss"
+  },
+  {
+    "outputFile": "wwwroot/scripts/todoTemplate.es5.js",
+    "inputFile": "wwwroot/scripts/todoTemplate.js"
+  },
+  {
+    "outputFile": "wwwroot/service-worker.published.js",
+    "inputFile": "wwwroot/service-worker.js"
+  }
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/compilerconfig.json:69]: plz correct the formatting in this file based on the configurations of the `.editorcofig` file. (tab-size & indenting-size equals to 4 spaces)

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs [hunk 5]
```diff
@@ -60,14 +63,49 @@ await HttpClient.PostAsJsonAsync("Auth/SignUp", new()
...
+
+            SignUpMessageType = BitMessageBarType.Success;
+
+            SignUpMessage = "The confirmation link has been re-sent.";
+        }
+        catch (KnownException e)
+        {
+            SignUpMessageType = BitMessageBarType.Error;
+
+            SignUpMessage = ErrorStrings.ResourceManager.Translate(e.Message, SignUpModel.Email!);
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs:91]: plz don't use `!` to bypass nullability checks, instead check at the start of the method that this property (`SignUpModel.Email`) is not null (or if it is null, simply return).

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor [hunk 6]
```diff
@@ -20,45 +20,67 @@
...
-            <ValidationMessage For="@(() => SignUpModel.Password)" />
-        </div>
 
-        <div>
-            <BitCheckbox Class="form-checkbox" @bind-Value="SignUpModel.IsAcceptPrivacy">
-                I agree to the <BitLink Href="/privacy" Target="_blank">Bit Privacy Policy</BitLink>
-            </BitCheckbox>
-            <ValidationMessage For="@(() => SignUpModel.IsAcceptPrivacy)" />
-        </div>
+        @if (IsSignedUp is false)
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor:24]: plz revert this if statement. it's better to check for the `true` value if there is an else block.

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 7]
```diff
@@ -47,36 +47,49 @@ IServer server
...
-                await _userManager.DeleteAsync(existingUserIfAny);
-                userToAdd.ConfirmationEmailRequestedOn = existingUserIfAny.ConfirmationEmailRequestedOn;
+                await _userManager.DeleteAsync(existingUser);
+                userToAdd.ConfirmationEmailRequestedOn = existingUser.ConfirmationEmailRequestedOn;
             }
         }
 
         var result = await _userManager.CreateAsync(userToAdd, signUpRequest.Password);
 
         if (!result.Succeeded)
```
Comment (Cyrus-Sushiant) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:69]: Plz, use `is false` instead of `!`

**Review by Cyrus-Sushiant — CHANGES_REQUESTED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor [hunk 8]
```diff
@@ -20,45 +20,67 @@
...
-            </BitCheckbox>
-            <ValidationMessage For="@(() => SignUpModel.IsAcceptPrivacy)" />
-        </div>
+        @if (IsSignedUp)
+        {
+            <BitButton Class="form-submit-button"
+                ButtonStyle="BitButtonStyle.Primary"
+                Title="Re-send link"
+                AriaLabel="Re-send link"
+                OnClick="WrapHandled(ResendLink)">
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor:30]: code formatting issue

**Review by msynk — CHANGES_REQUESTED** [review]

**Comment by msynk** [comment]
@maryamhdr & @Cyrus-Sushiant 
plz re-review this PR

**Review by msynk — DISMISSED** [review]

**Comment by msynk** [comment]
There are still some code formatting issues remains that I leave to the other reviewers 😁
@maryamhdr & @Cyrus-Sushiant

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 9]
```diff
@@ -47,36 +47,49 @@ IServer server
     [HttpPost("[action]")]
     public async Task SignUp(SignUpRequestDto signUpRequest, CancellationToken cancellationToken)
     {
-        var existingUserIfAny = await _userManager.FindByNameAsync(signUpRequest.UserName);
+        var existingUser = await _userManager.FindByNameAsync(signUpRequest.UserName);
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:51]: No blank line is needed here

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs [hunk 10]
```diff
@@ -47,36 +47,49 @@ IServer server
...
     [HttpPost("[action]")]
     public async Task SendConfirmationEmail(SendConfirmationEmailRequestDto sendConfirmationEmailRequest, CancellationToken cancellationToken)
     {
         var user = await _userManager.FindByEmailAsync(sendConfirmationEmailRequest.Email);
 
+        if (user is null)
+            throw new BadRequestException(nameof(ErrorStrings.UserNameNotFound));
+        
+        if (await _userManager.IsEmailConfirmedAsync(user))
+            throw new BadRequestException(nameof(ErrorStrings.EmailAlreadyConfirmed));
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Controllers/AuthController.cs:86]: I think it's better to wrap the statement of the `if` condition between `{...}`

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs [hunk 11]
```diff
@@ -6,6 +6,7 @@ public partial class SignUp
 {
     public SignUpRequestDto SignUpModel { get; set; } = new();
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs:8]: The blank line is not needed here

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs [hunk 12]
```diff
@@ -6,6 +6,7 @@ public partial class SignUp
...
 
+    public bool IsSignedUp { get; set; }
     public bool IsLoading { get; set; }
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs:11]: The blank line is not needed here

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs [hunk 13]
```diff
@@ -34,8 +35,10 @@ private async Task DoSignUp()
 
             await HttpClient.PostAsJsonAsync("Auth/SignUp", SignUpModel, TodoTemplateJsonContext.Default.SignUpRequestDto);
 
+            IsSignedUp = true;
+
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs:39]: The blank line is not needed here

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs [hunk 14]
```diff
@@ -56,14 +59,49 @@ private async Task DoSignUp()
...
+        try
+        {
+            await HttpClient.PostAsJsonAsync("Auth/SendConfirmationEmail", new()
+            {
+                Email = SignUpModel.Email
+            }, TodoTemplateJsonContext.Default.SendConfirmationEmailRequestDto);
+
+            SignUpMessageType = BitMessageBarType.Success;
+
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs:80]: The blank line is not needed here

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs [hunk 15]
```diff
@@ -56,14 +59,49 @@ private async Task DoSignUp()
...
+            SignUpMessage = "The confirmation link has been re-sent.";
+        }
+        catch (KnownException e)
+        {
+            SignUpMessageType = BitMessageBarType.Error;
+
```
Comment (maryamhdr) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Web/Pages/SignUp.razor.cs:86]: The blank line is not needed here

**Review by maryamhdr — CHANGES_REQUESTED** [review]

**Review by Cyrus-Sushiant — APPROVED** [review]

**Review by maryamhdr — APPROVED** [review]

**Review by Cyrus-Sushiant — APPROVED** [review]

**Review by msynk — APPROVED** [review]

**Pull request merged**